src/crimson/osd/osd_operations/client_request.cc

```diff
@@ -159,4 +163,29 @@ seastar::future<> ClientRequest::process_op(
   }));
 }
 
+bool ClientRequest::is_misdirected(const PG& pg) const
+{
+  // primary can handle both normal ops and balanced reads
+  if (pg.is_primary()) {
+    return false;
+  }
+  // otherwise take a closer look
+  if (const int flags = m->get_flags();
+      flags & CEPH_OSD_FLAG_BALANCE_READS ||
+      flags & CEPH_OSD_FLAG_LOCALIZE_READS) {
+    if (!op_info.may_read()) {
+      // no read found, so it can't be balanced read
+      return true;
+    }
+    if (op_info.may_write() || op_info.may_cache()) {
+      // write op, but i am not primary
+      return true;
+    }
+    // balanced reads; any replica will do
+    return pg.is_non_primary();
```

Comment:
```cpp
  /// acting osd that is not the primary
  bool is_nonprimary() const {
    return role >= 0 && pg_whoami != primary;
  }
```

Hmm, looks we really need to call this at the check is not just a negation of `is_primary()` we already verified above.